beacon-chain/state/v3/types.go

```diff
@@ -0,0 +1,76 @@
+package v3
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/fieldtrie"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/stateutil"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state/types"
+	"github.com/prysmaticlabs/prysm/config/params"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+)
+
+func init() {
+	fieldMap = make(map[types.FieldIndex]types.DataType, params.BeaconConfig().BeaconStateMergeFieldCount)
+
+	// Initialize the fixed sized arrays.
+	fieldMap[types.BlockRoots] = types.BasicArray
+	fieldMap[types.StateRoots] = types.BasicArray
+	fieldMap[types.RandaoMixes] = types.BasicArray
+
+	// Initialize the composite arrays.
+	fieldMap[types.Eth1DataVotes] = types.CompositeArray
+	fieldMap[types.Validators] = types.CompositeArray
+}
+
+// Field Aliases for values from the types package.
+const (
+	genesisTime                    = types.GenesisTime
+	genesisValidatorRoot           = types.GenesisValidatorRoot
+	slot                           = types.Slot
+	fork                           = types.Fork
+	latestBlockHeader              = types.LatestBlockHeader
+	blockRoots                     = types.BlockRoots
+	stateRoots                     = types.StateRoots
+	historicalRoots                = types.HistoricalRoots
+	eth1Data                       = types.Eth1Data
+	eth1DataVotes                  = types.Eth1DataVotes
+	eth1DepositIndex               = types.Eth1DepositIndex
+	validators                     = types.Validators
+	balances                       = types.Balances
+	randaoMixes                    = types.RandaoMixes
+	slashings                      = types.Slashings
+	previousEpochParticipationBits = types.PreviousEpochParticipationBits
+	currentEpochParticipationBits  = types.CurrentEpochParticipationBits
+	justificationBits              = types.JustificationBits
+	previousJustifiedCheckpoint    = types.PreviousJustifiedCheckpoint
+	currentJustifiedCheckpoint     = types.CurrentJustifiedCheckpoint
+	finalizedCheckpoint            = types.FinalizedCheckpoint
+	inactivityScores               = types.InactivityScores
+	currentSyncCommittee           = types.CurrentSyncCommittee
+	nextSyncCommittee              = types.NextSyncCommittee
+	latestExecutionPayloadHeader   = types.LatestExecutionPayloadHeader
+)
+
+// fieldMap keeps track of each field
+// to its corresponding data type.
+var fieldMap map[types.FieldIndex]types.DataType
+
+// ErrNilInnerState returns when the inner state is nil and no copy set or get
+// operations can be performed on state.
+var ErrNilInnerState = errors.New("nil inner state")
+
+// BeaconState defines a struct containing utilities for the eth2 chain state, defining
+// getters and setters for its respective values and helpful functions such as HashTreeRoot().
+type BeaconState struct {
+	state                 *ethpb.BeaconStateMerge
+	lock                  sync.RWMutex
+	dirtyFields           map[types.FieldIndex]interface{}
```

Comment:
This is `map[types.FieldIndex]bool` in v2, perhaps is better to uniformize them?

Reply:
+1 , also `map[types.FieldIndex]bool` avoids the inefficiencies that come with using an interface type.